### ISSUE TITLE
MASTER_FAILBACK_INTERVAL  not working

### ISSUE THREAD

**Issue by harmv** [saltstack/salt#39439]
### Description of Issue/Question

We are using MASTER_FAILBACK_INTERVAL  to detect a master failure and switch over.
This does not seem to be working as advertised.

It does seem to work if a master cleanly shuts down. But if a master becomes unreachable, because for example the network has a problem.  Then MASTER_FAILBACK_INTERVAL  does _not_ detect this, and the next master is not chosen.

### Setup

minion

     master:
        - 10.130.11.12
        - 10.130.11.253
    log_level: debug
    master_type: failover
    master_alive_interval: 10

### Steps to Reproduce Issue

I let the minion connect to the 10.130.11.12 master.

I monitor the minion traffic with tcpdump (both on master as on minion side)
 * minion:  `tcpdump "host 10.130.11.12 or host 10.130.11.253"`
 * master:  `tcpdump "host <ip of minion"`

I unplug the ethernet cable of the minion (in a switch further down stream, so the minion doesn't see an ethernet disconnect itself).

###  Expected behaviour
When the connection is ok, I expect to see some keepalive-kinda traffic with interval 10s (MASTER_FAILBACK_INTERVAL)  visible on the connection.

After 10 seconds MASTER_FAILBACK_INTERVAL  is notices that the master is not reachable. And it tries to establish a new connection to the second master.

### Actual  behaviour

When the connection is ok.  I don't see any traffic with interval of 10s.
I use tcpdump at the master (and on the minion side) to monitor the traffic.
If the minion starts I see traffic, after a while all settles down and _no_ traffic happens anymore, no bytes every 10s, nothing, nada)

So the minion _cannot_ detect if its connection with the master is ok. (it needs to talk to the master to know its up).

When I disconnecting the ethernet cable between the minion and the master.... the minion never knowns this happens, and *doesn't reconnect* to the 2nd master.

logging:
The minion happily logs stuff about __master_alive, however when looking at the network level, thats _not_ what is doing.(tcpdump).

    Feb 16 12:36:19 14100018 salt-minion[28642]: [DEBUG   ] schedule.handle_func: Removing /var/cache/salt/minion/proc/20170216123619698461
    Feb 16 12:36:29 14100018 salt-minion[28642]: [INFO    ] Running scheduled job: __master_alive
    Feb 16 12:36:29 14100018 salt-minion[28642]: [DEBUG   ] schedule: This job was scheduled with jid_include, adding to cache (jid_include defaults to True)
    Feb 16 12:36:29 14100018 salt-minion[28642]: [DEBUG   ] schedule: This job was scheduled with a max number of 1
    Feb 16 12:36:29 14100018 salt-minion[28642]: [DEBUG   ] schedule.handle_func: adding this job to the jobcache with data {'fun_args': [], 'jid': '20170216123629693268', 'schedule': '__master_alive', 'pid': 2515, 'fun': 'status.master', 'id': '14100018'}
    Feb 16 12:36:29 14100018 salt-minion[28642]: [DEBUG   ] schedule.handle_func: Removing /var/cache/salt/minion/proc/20170216123629693268
    Feb 16 12:36:39 14100018 salt-minion[28642]: [INFO    ] Running scheduled job: __master_alive
    Feb 16 12:36:39 14100018 salt-minion[28642]: [DEBUG   ] schedule: This job was scheduled with jid_include, adding to cache (jid_include defaults to True)
    Feb 16 12:36:39 14100018 salt-minion[28642]: [DEBUG   ] schedule: This job was scheduled with a max number of 1
    Feb 16 12:36:39 14100018 salt-minion[28642]: [DEBUG   ] schedule.handle_func: adding this job to the jobcache with data {'fun_args': [], 'jid': '20170216123639704620', 'schedule': '__master_alive', 'pid': 2560, 'fun': 'status.master', 'id': '14100018'}
    Feb 16 12:36:40 14100018 salt-minion[28642]: [DEBUG   ] schedule.handle_func: Removing /var/cache/salt/minion/proc/20170216123639704620

### Notes
similar to #38116?

I noticed a difference in `transport=tcp`  and zeromq (default)

* zeromq:   10s periodic logging as show above, but no periodic traffic to the master
* `transport=tcp`:    *no* periodic logging,  no periodic traffic to the master

### Versions Report
2016.11.2  but also seen on 2016.3.3

**Comment by Ch3LL** [comment]
### Expected Behavior
Okay glancing at the code it checks these files [here](https://github.com/saltstack/salt/blob/2016.11/salt/utils/network.py#L908) if the connection to the master is in that file then it does not send network traffic over the line as seen [here](https://github.com/saltstack/salt/blob/2016.11/salt/modules/status.py#L1063). It only will fire an event if the master is not in that list as shown [here](https://github.com/saltstack/salt/blob/2016.11/salt/modules/status.py#L1068) So if i read that code correctly I wouldn't expect to see network traffic every time the scheduled `status.master` job is run. Although I'm not the expert here. @DmitryKuzmenko did I interpret that correctly?

In my testing when the master and minion are both up and running and the network is okay I do see network traffic sometimes. Its not very often, i'm assuming that information is lost from that file when the tcp connection is no longer alive and so we send that event to re-establish the conneciton. Are you not seeing any traffic at all? I would give it a good 5-10 minute before you say no

### My test case
In my test setup I do not have a switch in between that I can just disconnect to test the other issue you have mentioned. Could you please add the following log message:

```
                    iret = _parse_tcp_line(line)
                    log.debug('IRET: {0}'.format(iret))
                    sl = next(iter(iret))
```

In the file `salt/utils/network.py` . You should then see when you break the network connection in between the master and the minion if that your master information stays in the output of `IRET` in the debug log files.

One test case thats semi like yours is I enabled the firewall on the master where the minion is currently connected to. When it attempts to fire the master status event I see this in tcpdump

```
12:47:33.773407 IP ch3ll-one-minion.52866 > ch3ll-one-master.4505: Flags [.], ack 1, win 229, options [nop,nop,TS val 4969984 ecr 4477420], length 0
12:47:33.774452 IP ch3ll-one-master > ch3ll-one-minion: ICMP host ch3ll-one-master unreachable - admin prohibited, length 60
```

Notice this; `unreachable - admin prohibited, length 60`. When it sees that network data it does not fail over to the master immediately. It takes about 15 -20 minutes before it realizes its not established anymore then fails over. So why isn't the minion seeing this as unreachable and NOT connecting to the next master. Although this could also be related to the firewall not killing already established connections and ICMP is a ping not a connection over 4505 so this might not be the best test case.

### Other possible issue
The other possible issue is you don't have that `/proc/net/tcp` file and its actually calling `_linux_remotes_on` and there's an issue in that function and it looks like it looks for the output of `lsof -iTCP:port -n -P`

### TLDR; 
  - I would expect not to see network traffic every time the master_alive job is scheduled if i read the code correctly. Can you check that you wait atleast 5-10 minutes to see if you don't see any network traffic when everything is okay and there are no network failures in between.
  - I need you to add some debug logs as detailed above for your exact use case and see if you still see your master host in the output of the debug log message when there is a network failure.

**Comment by Ch3LL** [comment]
Also for the tcp transport issues you are probably running into this: https://github.com/saltstack/salt/issues/39463

**Comment by harmv** [comment]
I don't think  checking /proc/net/tcp  is a reliable way to detect if a 'master is alive'.   I think `master_alive_interval` only makes sense if you actively poll the server with that interval. Thats the only way you _know_ for sure that the master is alive.  
problems with /proc/net/tcp:
 * you might miss network issues (as in the problem reported)
 * you might miss a totally overloaded server. (where salt-master no longer responses)
 * the minion might recover after after hours  (or faster)... but that kinda beats the purpose of having an `master_alive_interval` at  all.  Setting `master_alive_interval`, indicates you want *that* specific interval, not a random timeout the OS decides that a socket is removed from /proc/net/tcp

(Yes I _do_ have a /proc/net/tcp  file, its not falling back to `linux_remotes_on`)

What would make sense here, in my opinion, is that the minion calls `status.ping_master` to a connected master with interval `master_alive_interval` if `master_alive_interval` is set. `status.master` does not seem to be reliable enough for this.

**Comment by harmv** [comment]
As for the tcp transport issue, when looking in the  minion.py code, I see it actually skips all calls to`status.master` (or `status.ping_master` for that sake) when transport happens to be tcp.

minion.py
```
 if (self.opts['transport'] != 'tcp' and
                self.opts['master_alive_interval'] > 0 and
                self.connected):
...
           if self.opts['transport'] != 'tcp':
                    schedule = {
                       'function': 'status.master'
```

I don't understand why this is done. I'd say the `master_alive_interval` is as useful for transport tcp as for any other.  
(That is... if it actually verifies that the master is 'alive')
If its important for a setup, to fallback to 2nd master, if something is wrong. You need this also with transport tcp, no?

**Comment by harmv** [comment]
So. the good part.  (transport tcp)

When waiting longer, I _do_ see the minion detect the server beeing down.
After 10 minutes, it somehow (I fail to see how in the code, as the minion skips all `status.master` calls in minion.py).  detects the connection is lost, and it connects to the new server.


```
Feb 21 11:08:26 14100018 salt-minion[6348]: [DEBUG   ] LazyLoaded config.merge
.... nothing more logged
.... then at 11:19 the connection is phsically broken by me
... 10 min later... its noticed
Feb 21 11:29:42 14100018 salt-minion[6348]: [DEBUG   ] tcp stream to 10.130.11.12:4505 closed, unable to recv
Feb 21 11:29:42 14100018 salt-minion[6348]: [DEBUG   ] Sending event - data = {'_stamp': '2017-02-21T10:29:42.052631', 'master': '10.130.11.12'}
Feb 21 11:29:42 14100018 salt-minion[6348]: [DEBUG   ] Handling event tag '__master_disconnected'
Feb 21 11:29:42 14100018 salt-minion[6348]: [INFO    ] Connection to master 10.130.11.12 lost
Feb 21 11:29:42 14100018 salt-minion[6348]: [INFO    ] Trying to tune in to next master from master-list
Feb 21 11:29:42 14100018 salt-minion[6348]: [INFO    ] Moving possibly failed master 10.130.11.12 to the end of the list of masters
Feb 21 11:29:42 14100018 salt-minion[6348]: [DEBUG   ] Connecting to master. Attempt 1 of 1
Feb 21 11:29:42 14100018 salt-minion[6348]: [WARNING ] Master ip address changed from 10.130.11.12 to 10.130.11.253
```


So eventually its working, but the timeout set in `master_alive_interval`  is not obeyed/effective.

**Comment by harmv** [comment]
Hm...  I seem to have found the knobs.  At least  similar knobs...

`tcp_keepalive_idle`, `tcp_keepalive_cnt` & `tcp_keepalive_intvl` Seem to control exactly this.  (Only when using transport tcp !!)


When having network problems, the minion reconnects to the next master, in the timeout defined by `tcp_keepalive_*`.
With 
```
tcp_keepalive_idle = 30
tcp_keepalive_intvl = 10
tcp_keepalive_cnt = 3
```
I get a minion that reconnects in one minute, if the master looses its network.


This is kinda confusing from the documentation  though...   `master_alive_interval`   seems to be intended for this too.  As far as I see, `master_alive_interval` is not doing anything useful now. (it misses cases its intended for, it only works if a master cleanly shuts down).

Note that the above only works on `transport=tcp`, `tcp_keepalive_*` does not seem to be working when using `transport=zmq`.  (The keealives are never sent by the zeromq library)

**Comment by Ch3LL** [comment]
@harmv thanks for all the great work here. Since this is a design decision i would like to get some more ideas on this one.

ping @DmitryKuzmenko @cachedout or @thatch45 

If you see my previous comment you can see that `status.master` checks the file `/proc/net/tcp` to check to see if the master is still connected. (if i interpreted the code correctly) And according to @harmv 's issue there seems to be certain situations were the connection stays in this file even though there are network failures. So my question is there a better way to detect connection between the master/minion? Or do we just simply need to add better documentation stating that one can add those tcp_keepalive settings to ensure network failures are accounted for quickly?

**Comment by riq-dfaubion** [comment]
I wish I could understand why the devs are so resistant to application level keepalives.  Seems it would solve so many of these issues.
